javaslang/src/main/java/javaslang/collection/TreeMap.java

```diff
@@ -477,6 +505,98 @@ public boolean isEmpty() {
         return TreeSet.ofAll(keyComparator(), iterator().map(Tuple2::_1));
     }
 
+    @Override
+    public Iterator<Tuple2<K, V>> descendingIterator() {
+        return entries.descendingIterator();
+    }
+
+    @Override
+    public SortedSet<K> descendingKeySet() {
+        return descendingMap().keySet();
+    }
+
+    @Override
+    public SortedMap<K, V> descendingMap() {
+        return new TreeMap<>(entries.descendingTree());
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(Option<K> fromKey, boolean fromInclusive, Option<K> toKey, boolean toInclusive) {
+        return new TreeMap<>(
+                entries.subTree(
+                    fromKey.map(k -> Tuple.of(k, null)), fromInclusive,
+                    toKey.map(k -> Tuple.of(k, null)), toInclusive));
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, boolean fromInclusive, K toKey, boolean toInclusive) {
+        return subMap(Option.some(fromKey), fromInclusive, Option.some(toKey), toInclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> subMap(K fromKey, K toKey) {
+        return subMap(fromKey, true, toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey, boolean inclusive) {
+        return subMap(Option.none(), true, Option.some(toKey), inclusive);
+    }
+
+    @Override
+    public SortedMap<K, V> headMap(K toKey) {
+        return headMap(toKey, false);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey, boolean inclusive) {
+        return subMap(Option.some(fromKey), inclusive, Option.none(), true);
+    }
+
+    @Override
+    public SortedMap<K, V> tailMap(K fromKey) {
+        return tailMap(fromKey, true);
+    }
+
+    @Override
+    public Option<Tuple2<K, V>> floor(K key) {
+        return entries.floor(Tuple.of(key, null));
+    }
+
+    @Override
+    public Option<K> floorKey(K key) {
+        return floor(key).map(e -> e._1);
+    }
+
+    @Override
+    public Option<Tuple2<K, V>> ceiling(K key) {
+        return entries.ceiling(Tuple.of(key, null));
```

Comment:
Because we don't have the value ;)

The input parameter is a `key` and the output parameter is the key/value pair that meets the floor condition: highest key lower or equal than the key provided.

It will be easier to understand if you look at the `get(K key)` implementation. It receives a key as parameter and returns the key/value pair for that key. It is the same idea for the other methods:
```
    @Override
    public Option<V> get(K key) {
        final V ignored = null;
        return entries.find(new Tuple2<>(key, ignored)).map(Tuple2::_2);
    }
```

Also remember that `TreeMap` is just a thin wrapper around `RedBlackTree`.